partiql-parser/src/scanner.rs

```diff
@@ -238,371 +263,281 @@ mod test {
-                }
-            }),
-            Ok(Token {
-                content: Content::DecimalLiteral(BigDecimal::from_str_radix("0.0", 10).unwrap()),
-                start: LineAndColumn::at(1, 23),
-                end: LineAndColumn::at(1, 30),
-                text: "0.0e000",
-                remainder: Remainder {
-                    input: "",
-                    offset: LineAndColumn::at(1, 30)
-                }
-            }),
-            syntax_error("IGNORED MESSAGE", Position::at(1, 30)),
+        scanner_test_case![
+            "0" => integer_literal(0),
+            " ",
+            "000" => integer_literal(0),
+            " ",
+            ".0" => decimal_literal_from_str("0.0"),
+            " ",
+            "000.000" => decimal_literal_from_str("0.000"),
+            " ",
+            ".0e0" => decimal_literal_from_str("0.0"),
+            " ",
+            "0.0e000" => decimal_literal_from_str("0.0")
         ]
     )]
     #[case::select_from(
-        r#"SelEct '✨✨✨' fROM "┬─┬" "#,
-        vec![
-            Ok(Token {
-                content: Content::Keyword("SELECT".into()),
-                start: LineAndColumn::at(1, 1),
-                end: LineAndColumn::at(1, 7),
-                text: "SelEct",
-                remainder: Remainder {
-                    input: r#" '✨✨✨' fROM "┬─┬" "#,
-                    offset: LineAndColumn::at(1, 7)
-                }
-            }),
-            Ok(Token {
-                content: Content::StringLiteral("✨✨✨".into()),
-                start: LineAndColumn::at(1, 8),
-                end: LineAndColumn::at(1, 13),
-                text: "'✨✨✨'",
-                remainder: Remainder {
-                    input: r#" fROM "┬─┬" "#,
-                    offset: LineAndColumn::at(1, 13)
-                }
-            }),
-            Ok(Token {
-                content: Content::Keyword("FROM".into()),
-                start: LineAndColumn::at(1, 14),
-                end: LineAndColumn::at(1, 18),
-                text: "fROM",
-                remainder: Remainder {
-                    input: r#" "┬─┬" "#,
-                    offset: LineAndColumn::at(1, 18)
-                }
-            }),
-            Ok(Token {
-                content: Content::Identifier("┬─┬".into()),
-                start: LineAndColumn::at(1, 19),
-                end: LineAndColumn::at(1, 24),
-                text: r#""┬─┬""#,
-                remainder: Remainder {
-                    input: " ",
-                    offset: LineAndColumn::at(1, 24)
-                }
-            }),
-            syntax_error("IGNORED MESSAGE", Position::at(1, 25)),
+        scanner_test_case![
+            "SelEct" => keyword("SELECT"),
+            " ",
+            "'✨✨✨'" => string_literal("✨✨✨"),
+            " ",
+            "fROM" => keyword("FROM"),
+            " ",
+            r#""┬─┬""# => identifier("┬─┬"),
+            " "
```

Comment:
(Picking one case at random.)

What is the value in specifying the input *and* output (stuff on the left?) e.g.

```
"'✨✨✨'" => string_literal("✨✨✨"),
```

the plain literals make sense to me, but the pairs don't.

Like, for example, why not just have:

```
string_literal("✨✨✨"),
```

and, in the macro, you can transform that to the pairs. Nice thing about that is you could generate multiple test cases, e.g. adding whitepace or changing casing or something..

Reply:
btw. not saying you need to make your fancy macro even more fancy. just asking if I'm missing some specific point to the apparent duplication.

Reply:
The main reason why I didn't add that sophistication is I didn't want to start auto-generating lexical syntax--I still wanted that to be driven by the test writer.  Specifically, there are different ways to write things like identifiers so that could get complicated fast.  For other literals like integers, the lexical content could be very different than the logical form, and I wanted that to be easy to test (all the zero pad stuff for example).